src/bitcoin/electrum/mod.rs

```diff
@@ -0,0 +1,259 @@
+    GenesisHashMismatch(
+        BlockHash, /*expected hash*/
+        BlockHash, /*actual hash*/
+    ),
+    BdkWallet(String),
+}
+
+impl std::fmt::Display for ElectrumError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            ElectrumError::Server(e) => write!(f, "Electrum error: '{}'.", e),
+            ElectrumError::GenesisHashMismatch(expected_hash, actual_hash) => {
+                write!(
+                    f,
+                    "Genesis hash mismatch. The genesis hash is expected to be '{}' but was found to be '{}'.",
+                    expected_hash, actual_hash
+                )
+            }
+            ElectrumError::BdkWallet(e) => write!(f, "BDK wallet error: '{}'.", e),
+        }
+    }
+}
+
+/// Interface for Electrum backend.
+pub struct Electrum {
+    client: ElectrumClient,
+    bdk_wallet: wallet::BdkWallet,
+}
+
+impl Electrum {
+    pub fn new(
+        electrum_config: &config::ElectrumConfig,
+        network: &bitcoin::Network,
+        main_descriptor: &LianaDescriptor,
+    ) -> Result<Self, ElectrumError> {
+        let client = ElectrumClient::new(&electrum_config.addr.to_string(), network)?;
+        let genesis_hash = client.genesis_block().hash;
+        let bdk_wallet = wallet::BdkWallet::new(main_descriptor, genesis_hash);
+
+        Ok(Self { client, bdk_wallet })
+    }
+
+    pub fn client(&self) -> &ElectrumClient {
+        &self.client
+    }
+
+    pub fn wallet_coins(&self, outpoints: Option<&[OutPoint]>) -> HashMap<OutPoint, Coin> {
+        self.bdk_wallet.coins(outpoints)
+    }
+
+    pub fn init_wallet(
+        &mut self,
+        tip: &BlockChainTip,
+        coins: &[Coin],
+        txs: &[bitcoin::Transaction],
+        receive_index: ChildNumber,
+        change_index: ChildNumber,
+    ) {
+        self.bdk_wallet
+            .init(tip, coins, txs, receive_index, change_index)
+    }
+
+    pub fn sync_wallet(
+        &mut self,
+        receive_index: ChildNumber,
+        change_index: ChildNumber,
+    ) -> Result<(), ElectrumError> {
+        self.bdk_wallet
+            .sync_with_electrum(&self.client.0, receive_index, change_index)
+            .map_err(ElectrumError::Server)
+    }
+
+    pub fn common_ancestor(&self, tip: &BlockChainTip) -> Option<BlockChainTip> {
+        self.bdk_wallet
+            .common_ancestor_with_electrum(&self.client.0, tip)
+    }
+
+    pub fn wallet_transaction(
+        &self,
+        txid: &bitcoin::Txid,
+    ) -> Option<(bitcoin::Transaction, Option<Block>)> {
+        self.bdk_wallet.get_transaction(txid)
+    }
+}
+
+// If Electrum takes more than 3 minutes to answer one of our queries, fail.
+const RPC_SOCKET_TIMEOUT: u8 = 180;
+
+// Number of retries while communicating with the Electrum server.
+// A retry happens with exponential back-off (base 2) so this makes us give up after (1+2+4+8+16+32=) 63 seconds.
+const RETRY_LIMIT: u8 = 6;
+
+pub struct ElectrumClient(Client);
```

Comment:
I think I should put this `ElectrumClient` in a separate "electrum/client.rs" file. Then the utils functions `block_hash` and `chain_tip` could be part of this client (I added separate utils functions for these in order to avoid circular dependencies between wallet.rs and mod.rs).

Reply:
I've now moved the Electrum client logic to client.rs and moved the sync logic out of wallet.rs and to the `Electrum` interface as it combines both the client and wallet.

Reply:
Note the wallet.rs file has no Electrum-specific logic and could be used for other backend types in future. At some point we might like to move it out of the electrum/ directory.